src/grain/distribution.js

```diff
@@ -0,0 +1,232 @@
+// @flow
+
+/**
+ * In SourceCred, projects regularly distribute Grain to contributors based on
+ * their Cred scores. This is called a "Distribution".
+ *
+ * This module contains the logic for calculating distribution amounts for
+ * contributors. A distribution contains "receipts" showing how much Grain each
+ * contributor will receive, the timestamp of the distribution in question, and the
+ * "strategy" used to distribute grain.
+ *
+ * Currently we support two strategies:
+ * - IMMEDIATE, which distributes a fixed amount of grain based on the cred scores
+ * in the most recent completed time interval
+ * - LIFETIME, which distributes a fixed amount of grain based on cred scores
+ * across
+ * all time, prioritizing paying people who were under-paid historically (i.e.
+ * their lifetime earnings are lower than we would expect given their current
+ * cred score)
+ *
+ * In all cases, the timestamp for the distribution is used to determine which cred
+ * scores are in scope. For example, if you create a immediate distribution with a
+ * timestamp in the past, it will reward people with cred in the past time
+ * period, not the current one.
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {type Grain, multiplyFloat, toFloatRatio, ZERO} from "./grain";
+
+export const DISTRIBUTION_VERSION_1 = 1;
+
+export type DistributionStrategy = ImmediateV1 | LifetimeV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type LifetimeV1 = {|
+  +type: "LIFETIME",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type GrainReceipt = {|+address: NodeAddressT, amount: Grain|};
+
+export type DistributionV1 = {|
+  +type: "DISTRIBUTION",
+  +timestampMs: number,
+  +version: number,
+  +strategy: DistributionStrategy,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
+
+export type CredTimeSlice = {|
+  +intervalEndMs: number,
+  +cred: Map<NodeAddressT, number>,
+|};
+
+export type CredHistory = $ReadOnlyArray<CredTimeSlice>;
```

Comment:
I'm not very fond of the TimelineCred class, and I expect that work on CredRank will leave us with something different. That's why I like the idea of having a simple type for representing cred here, rather than coupling it to TimelineCred--it minimizes the churn in the future when we replace TimelineCred.

Reply:
> I'm not very fond of the TimelineCred class, and I expect that work on CredRank will leave us with something different.

That's fine, because we're using `Map<NodeAddressT, number>` anyway, so we can just replace these functions I'm suggesting with ones that take CredRank data instead:

```js
function latestCred(
  credRankResult: CredRankResult,
  userPrefixes: $ReadOnlyArray<NodeAddressT>,
  intervalEndMs: number
): Map<NodeAddressT, number> {}

function lifetimeCred(
  credRankResult: CredRankResult,
  userPrefixes: $ReadOnlyArray<NodeAddressT>,
  intervalEndMs: number
): Map<NodeAddressT, number> {}
```

But now we're creating a new _Cred_ representation and making it use-case specific by placing it in `grain/distribution` as an indirection to not have to deal with `TimelineCred`, while not producing instances of this type here, and don't need some of the data it contains, ... well I'm confused :stuck_out_tongue: and not sure this is an improvement over `TimelineCred`.

Reply:
You make a good point, let's follow up on it once we're actually building the interface to TimelineCred. Even if we make a suboptimal decision here, it stays super local and is easy to fix later, so I'm going to mark this resolved so we can keep momentum on this PR.

Reply:
Tracking at https://github.com/sourcecred/sourcecred/issues/1759